lib/gssapi/krb5/name_attrs.c

```diff
@@ -0,0 +1,691 @@
+        *minor_status = kret;
+        return kret == 0 ? GSS_S_COMPLETE : GSS_S_FAILURE;
+    }
+
+    return GSS_S_UNAVAILABLE;
+
+enomem:
+    if (value) {
+        free(value->value);
+        value->length = 0;
+        value->value = 0;
+    }
+    *minor_status = ENOMEM;
+    return GSS_S_FAILURE;
+}
+
+static OM_uint32
+add_urn(OM_uint32 *minor_status,
+        gss_name_t name,
+        gss_buffer_t urn,
+        gss_buffer_set_t *attrs)
+{
+    OM_uint32 major;
+
+    major = _gsskrb5_get_name_attribute(minor_status, name, urn,
+                                        0, 0, 0, 0, 0);
+    if (major == GSS_S_COMPLETE) {
+        major = gss_add_buffer_set_member(minor_status, urn, attrs);
+        if (major)
+            return major;
+    }
+    if (major == GSS_S_UNAVAILABLE)
+        return GSS_S_COMPLETE;
+    return major;
+}
+
+#define ADD_URN(l)                                                      \
+    do if (major == GSS_S_COMPLETE) {                                   \
+        if (strncmp(l, "urn:", sizeof("urn:") - 1) == 0) {              \
+            urn.value = l;                                              \
+            urn.length = sizeof(l) - 1;                                 \
+        } else {                                                        \
+            urn.value = GSS_KRB5_NAME_ATTRIBUTE_BASE_URN l;             \
+            urn.length = sizeof(GSS_KRB5_NAME_ATTRIBUTE_BASE_URN l) - 1;\
+        }                                                               \
+        major = add_urn(minor_status, name, &urn, attrs);               \
+    } while (0)
+
+OM_uint32 GSSAPI_CALLCONV
+_gsskrb5_inquire_name(OM_uint32 *minor_status,
+                      gss_name_t name,
+                      int *name_is_MN,
+                      gss_OID *MN_mech,
+                      gss_buffer_set_t *attrs)
+{
+    OM_uint32 major = GSS_S_COMPLETE;
+    gss_buffer_desc urn;
+    krb5_error_code ret;
+    krb5_context context;
+    char lname[32];
+
+    GSSAPI_KRB5_INIT(&context);
+
+    *minor_status = 0;
+    if (name_is_MN)
+        *name_is_MN = 1;
+    if (MN_mech)
+        *MN_mech = GSS_KRB5_MECHANISM;
+    if (name == GSS_C_NO_NAME)
+        return GSS_S_CALL_INACCESSIBLE_READ;
+    if (attrs == NULL)
+        return GSS_S_CALL_INACCESSIBLE_WRITE;
+    ADD_URN("realm");
+    ADD_URN("name-ncomp");
+    ADD_URN("name-ncomp#0");
+    ADD_URN("name-ncomp#1");
+    ADD_URN("name-ncomp#2");
+    ADD_URN("name-ncomp#3");
+    ADD_URN("name-ncomp#4");
+    ADD_URN("name-ncomp#5");
+    ADD_URN("name-ncomp#6");
+    ADD_URN("name-ncomp#7");
+    ADD_URN("name-ncomp#8");
+    ADD_URN("name-ncomp#9");
+    ADD_URN("peer-realm");
+    ADD_URN("ticket-authz-data#pac");
+    ADD_URN("urn:mspac:");
+    ADD_URN("authenticator-authz-data"); /* XXX Add fragments? */
+    ADD_URN("ticket-authz-data"); /* XXX Add fragments? */
+    ADD_URN("authz-data");
+    ADD_URN("transit-path");
+    ADD_URN("canonical-name");
```

Comment:
could be part of table-driven approach?

Reply:
Only if I also do that for the get-name-attribute side.  I don't think it would be a simplification, but maybe that's just me.  I could do it though, if you feel strongly about it.